### PR TITLE
Infer delegate type from lambda expression with explicit return statements without expressions

### LINKED ISSUE #55320: Cannot infer delegate type for lambda expression with implicit and explicit `return` statements
Compile the following with `-langversion:10`:
```csharp
class Program
{
    static void Main()
    {
        F(true);
    }

    static System.Delegate F(bool b)
    {
        return () =>
        {
            if (b) return;
        };
    }
}
```

Expected:
The delegate type `System.Action` is inferred for the lambda expression.

Actual:
```
(10,16): error CS8917: The delegate type could not be inferred.
```

Relates to test plan https://github.com/dotnet/roslyn/issues/52192

### PR DESCRIPTION
**Pull request by cston**
Fixes #55320.

Test plan https://github.com/dotnet/roslyn/issues/52192

### CODE REVIEW

File: src/Compilers/CSharp/Portable/BoundTree/UnboundLambda.cs [hunk 1]
```diff
@@ -624,7 +624,7 @@ private static TypeWithAnnotations DelegateReturnTypeWithAnnotations(MethodSymbo
                 returnType = inferredReturnType.TypeWithAnnotations;
                 returnRefKind = inferredReturnType.RefKind;
 
-                if (!returnType.HasType && returnTypes.Count > 0)
+                if (!returnType.HasType && inferredReturnType.NumExpressions > 0)
```
Comment (AlekseyTs) [inline src/Compilers/CSharp/Portable/BoundTree/UnboundLambda.cs:627]: >if (!returnType.HasType && inferredReturnType.NumExpressions > 0) [](http://example.com/codeflow?start=16&length=65)

It looks like the change only affects scenarios that involve return statements without expressions. Is this correct? If so, consider adjusting commit's title/description as right now it feels too broad.

Comment (AlekseyTs) [inline src/Compilers/CSharp/Portable/BoundTree/UnboundLambda.cs:627]: >inferredReturnType.NumExpressions > 0 [](http://example.com/codeflow?start=43&length=37)

It looks like all added tests are success tests. Do we have failure tests for this condition? Specifically when there is a mix of returns with and without an expression.

**Comment by AlekseyTs** [comment]
Done with review pass (commit 1)

**Review by AlekseyTs — APPROVED** [review]
LGTM (commit 2)

**Review by 333fred — APPROVED** [review]
LGTM (commit 2)

**Pull request merged**